woocommerce/woocommerce-gutenberg-products-block
# Add Checkout form validation

Part of #1939.
Fixes #1848.
Fixes #1994.

This PR extends the validation context introduced in #1972 with some new methods:

* `hideValidationError` and `showValidationError` allow marking an error as visible or hidden to the user.
* `showAllValidationErrors` does the same but with all errors.
* `areThereValidationErrors` returns true whether there is at least one error.

It also changes the shape of the stored errors to `{ message: 'Test error', hidden: false }`. This allows adding errors in the context but displaying them conditionally. This way we can create validation errors when an input field is mounted, so if the user tries to submit the form without interacting with any field, those errors already exist and we can prevent the submit event and display the errors.

There are two new components that take care of creating validation errors on mount:
* `ValidatedTextInput` and `ValidatedSelect`, those components extend the `TextInput` and `Select` component we were currently using, but add validation on top of them. They also take care of clearing/hiding the errors when values are updated, the input is blurred, etc.

`ValidatedTextInput` relies on the native browser `<input>` validation rules. This has some advantages:
* We get email validation for free for the email input, required fields, etc. If we ever wanted to add new rules (min size for the address field, for example), we could do that with plain HTML.
* When a form fails to submit because there are validation errors, we can directly scroll to and focus the first invalid input.

That doesn't mean we can't build more complex validation rules. The validation check introduced in #1972 that verified a country was set when filling the address form in the shipping calculator, still works (and it works in the _Checkout_ block too).

### Screenshots

![Peek 2020-03-18 17-11](https://user-images.githubusercontent.com/3616980/76982049-9e6aae80-693b-11ea-89f0-78c8d127593f.gif)

### How to test the changes in this Pull Request:

1. Create a page with the _Checkout_ block and try filling the form leaving some fields empty, entering wrong emails in the email field, submitting the form without interacting with any field, etc.
2. Try entering the city, state or postal code before entering the Country and verify an error appears (only in the shipping form, it shouldn't happen in the billing form!).
3. Verify the coupon input field and the shipping calculator of the _Cart_ block still report errors properly like in #1972.
